### PR TITLE
[OPPRO-139] Enable BHJ

### PR DESCRIPTION
**Pull request by marin-ma**

### CODE REVIEW

File: jvm/src/main/scala/org/apache/spark/sql/execution/joins/BuildSideRelation.scala [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+  /**
+   * Deserialized relation from broadcasted value
+   */
+  def deserialized: Iterator[ColumnarBatch]
+
+  /**
+   * Returns a read-only copy of this, to be safely used in current thread.
+   */
+  def asReadOnlyCopy(): BuildSideRelation
```
Comment (marin-ma) [inline jvm/src/main/scala/org/apache/spark/sql/execution/joins/BuildSideRelation.scala:32]: @zzcclp BHJ basically reuses the implementation of SHJ, but the build side is replaced with broadcast relation. You may implement your `BuildSideRelation` to control the behavior of the broadcasted value. For example, make the `asReadOnlyCopy` return valid value for one thread and return empty value for other threads, and in the backend the thread holds valid value will be responsible for building hash table.

**Comment by zhouyuan** [comment]
FYI.
https://facebookincubator.github.io/velox/develop/joins.html#dynamic-filter-pushdown

**Comment by marin-ma** [comment]
It fails with AQE enabled. I'm working on the fix.

**Comment by marin-ma** [comment]
Verified on Jenkins. Q20 failed with AQE enabled has been fixed.

**Comment by rui-mo** [comment]
Verified locally on TPC-H q16. BHJ can be offloaded into native.

**Pull request merged**